ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -140,20 +145,103 @@ class GetBlobOperation extends GetOperation {
    * @param cryptoJobHandler {@link CryptoJobHandler} to assist in the execution of crypto jobs
    * @param time the Time instance to use.
    * @param isEncrypted if the encrypted bit is set based on the original blobId string of a {@link BlobId}.
+   * @param blobMetadataCache A cache to save blob metadata for composite blobs
    */
   GetBlobOperation(RouterConfig routerConfig, NonBlockingRouterMetrics routerMetrics, ClusterMap clusterMap,
       ResponseHandler responseHandler, BlobId blobId, GetBlobOptionsInternal options,
       Callback<GetBlobResultInternal> callback, RouterCallback routerCallback, BlobIdFactory blobIdFactory,
       KeyManagementService kms, CryptoService cryptoService, CryptoJobHandler cryptoJobHandler, Time time,
-      boolean isEncrypted, QuotaChargeCallback quotaChargeCallback) {
+      boolean isEncrypted, QuotaChargeCallback quotaChargeCallback, AmbryCache blobMetadataCache) {
     super(routerConfig, routerMetrics, clusterMap, responseHandler, blobId, options, callback, kms, cryptoService,
         cryptoJobHandler, time, isEncrypted);
     this.routerCallback = routerCallback;
     this.blobIdFactory = blobIdFactory;
     this.quotaChargeCallback = quotaChargeCallback;
-    firstChunk = new FirstGetChunk();
+    this.blobMetadataCache = blobMetadataCache;
+    if (shouldLookupMetadataCache()) {
+      BlobMetadata blobMetadata = (BlobMetadata) blobMetadataCache.getObject(blobId.toString());
+      if (blobMetadata != null) {
+        firstChunk = new CachedFirstChunk(blobMetadata);
+        blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheHit;
+      } else {
+        firstChunk = new FirstGetChunk();
+        blobMetadataCacheEntryStatus = AmbryCacheEntryStatus.cacheMiss;
+      }
+    } else {
+      firstChunk = new FirstGetChunk();
+    }
+  }
+
+  /**
+   * Decides if we should look up the metadata cache during a GET request
+   * @return True if and only if
+   *  - cache exists
+   *  - it is a RANGE request
+   *  - it is not a raw mode request
+   *  - the blobID is for a metadata blob
+   *  False, otherwise.
+   */
+  boolean shouldLookupMetadataCache() {
+    return (blobMetadataCache != null &&
+        options.getBlobOptions.getRange() != null &&
```

Comment:
Should we enable this for segmented GETs too (`options.getBlobOptions.hasBlobSegmentIdx()`) where user can request for segments of a stitched blob via URI like <blobId>/Segment/<SegmentNumber>? It can be useful for that feature too. Looks like you are already handling initializing data chunks for this case in `CachedFirstChunk::requestDataChunks()`

Reply:
I need to understand segmented GETs

Reply:
ok updated code to use cache for segmented GETs also